### PR TITLE
Core: Error on explicit actions while rendering or playing

### PR DESCRIPTION
**Pull request by kasperpeulen**
## What I did

* Previously we transformed the type of action args (void functions) to be optional, as a default value is injected based on a regex. For 8.0 we want to go to a more explicit direction, as it is impossible to filter them out correctly based on a regex. If action args are required for rendering, they should be explicitly added to the args. I removed the logic that turned implicit actions into spies. If an arg is needed in the play function as spy, it should also be explicitly added.
* The play function fails now when any error is thrown during the play function, even if it is thrown in the play function itself. This behavior is mimicked by how vitest works, and I also used the same flag `dangerouslyIgnoreUnhandledErrors` to disable this behavior:
image
* The above is necessary, to make sure that action can not be used during rendering or playing of the function. We now properly error out when someone runs the event during rendering or playing. This change is neccesary to make sure that tests behave the same if storybook is build with or without the `--test` flag.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-25238-sha-829df7a0`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25238-sha-829df7a0). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-25238-sha-829df7a0`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25238-sha-829df7a0) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`kasper/remove-ts-magic`](https://github.com/storybookjs/storybook/tree/kasper/remove-ts-magic) |
| **Commit** | [`829df7a0`](https://github.com/storybookjs/storybook/commit/829df7a0c5ae15b5f0b2c462c1f34a72671c5fb9) |
| **Datetime** | Fri Dec 29 17:55:22 UTC 2023 (`1703872522`) |
| **Workflow run** | [7358741213](https://github.com/storybookjs/storybook/actions/runs/7358741213) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=25238`_
</details>

### CODE REVIEW

**Review by tmeasday — COMMENTED** [review]
Does this require `MIGRATION.md` changes?

**Review by tmeasday — APPROVED** [review]
I'm surprised we don't have to remove any tests as a result of this :/ Maybe we should add some tests that at least exercise the new way of doing it and check it works? (Could be separate to this PR of course)

File: scripts/release/__tests__/label-patches.test.ts [hunk 1]
```diff
@@ -138,6 +138,7 @@ it('should label the PR associated with cherry picks in the current branch', asy
             .trim()
         : text
     )
+    // eslint-disable-next-line @typescript-eslint/no-shadow
     .filter((it) => it !== '');
```
Comment (yannbf) [inline scripts/release/__tests__/label-patches.test.ts:142]: it's best to change it so if we ever need to do a global find/replace from vitest's `it` to something else, it wouldn't affect this
```suggestion
    .filter((text) => text !== '');
```

File: code/renderers/vue3/template/cli/ts-3-8/Button.stories.ts [hunk 2]
```diff
@@ -11,9 +11,8 @@ const meta: Meta<typeof Button> = {
   argTypes: {
     size: { control: 'select', options: ['small', 'medium', 'large'] },
     backgroundColor: { control: 'color' },
-    onClick: { action: 'clicked' },
   },
-  args: { primary: false }, // default value
+  args: { primary: false, onClick: fn() }, // default value
```
Comment (yannbf) [inline code/renderers/vue3/template/cli/ts-3-8/Button.stories.ts:15]:
```suggestion
  args: {
    primary: false,
    onClick: fn()
  },
```

File: code/renderers/svelte/template/cli/ts-4-9/Button.stories.ts [hunk 3]
```diff
@@ -1,5 +1,4 @@
 import type { Meta, StoryObj } from '@storybook/svelte';
-
```
Comment (yannbf) [inline code/renderers/svelte/template/cli/ts-4-9/Button.stories.ts:2]: is this missing the onClick fn like the other templates are using?

File: code/renderers/svelte/template/cli/ts-3-8/Button.stories.ts [hunk 4]
```diff
@@ -1,5 +1,4 @@
 import type { Meta, StoryObj } from '@storybook/svelte';
-
```
Comment (yannbf) [inline code/renderers/svelte/template/cli/ts-3-8/Button.stories.ts:2]: is this missing the onClick fn like the other templates are using?

File: code/renderers/svelte/src/public-types.ts [hunk 5]
```diff
@@ -41,7 +41,7 @@ export type StoryFn<TCmpOrArgs = Args> = TCmpOrArgs extends SvelteComponentTyped
  * @see [Named Story exports](https://storybook.js.org/docs/formats/component-story-format/#named-story-exports)
  */
 export type StoryObj<MetaOrCmpOrArgs = Args> = MetaOrCmpOrArgs extends {
-  render?: ArgsStoryFn<SvelteRenderer, any>;
+  render?: ArgsStoryFn<any, any>;
```
Comment (yannbf) [inline code/renderers/svelte/src/public-types.ts:44]: is this intentional? is SvelteRenderer not correct anymore?

File: code/renderers/react/template/cli/ts-4-9/Button.stories.ts [hunk 6]
```diff
@@ -16,6 +16,7 @@ const meta = {
   argTypes: {
     backgroundColor: { control: 'color' },
   },
+  args: { onClick: fn() },
```
Comment (yannbf) [inline code/renderers/react/template/cli/ts-4-9/Button.stories.ts:19]: I think this needs proper explanation with a link to a tutorial/snippet, like it's done above for argTypes, tags, etc.

This applies for every CLI template

File: code/addons/interactions/template/stories/unhandled-errors.stories.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  component: globalThis.Components.Button,
+  args: {
+    label: 'Button',
+  },
+  parameters: {
+    actions: { argTypesRegex: '^on[A-Z].*' },
+  },
+};
+
+export const Default = {
```
Comment (yannbf) [inline code/addons/interactions/template/stories/unhandled-errors.stories.ts:14]: do you need to disable chromatic snapshots for this?

File: code/addons/interactions/template/stories/basics.stories.ts [hunk 8]
```diff
@@ -101,15 +102,14 @@ export const UserEventSetup = {
     const { args, canvasElement, step } = context;
     const user = userEvent.setup();
     const canvas = within(canvasElement);
-    await step('Select, type and paste on input using user-event v14 setup', async () => {
-      const input = await canvas.getByRole('textbox');
+    await step('Select and type on input using user-event v14 setup', async () => {
+      const input = canvas.getByRole('textbox');
       await user.click(input);
-      await user.type(input, 'Pasting: ');
-      await user.paste('foobar');
```
Comment (yannbf) [inline code/addons/interactions/template/stories/basics.stories.ts:108]: why was this removed?

File: code/addons/interactions/src/components/InteractionsPanel.tsx [hunk 9]
```diff
@@ -119,20 +126,34 @@ export const InteractionsPanel: React.FC<InteractionsPanelProps> = React.memo(
...
         )}
+        {unhandledErrors && (
+          <CaughtException>
+            <CaughtExceptionTitle>Unhandled Errors</CaughtExceptionTitle>
+            <CaughtExceptionDescription>
+              Found {unhandledErrors.length} unhandled error{unhandledErrors.length > 1 ? 's' : ''}{' '}
+              while running the play function. This might cause false positive assertions. Resolve
+              unhandled errors or ignore unhandled errors with setting the
+              <CaughtExceptionCode>test.dangerouslyIgnoreUnhandledErrors</CaughtExceptionCode>{' '}
+              parameter to <CaughtExceptionCode>true</CaughtExceptionCode>.
```
Comment (yannbf) [inline code/addons/interactions/src/components/InteractionsPanel.tsx:147]: I wonder if it would be better to point people to a link in the docs instead, which could explain things better with a code snippet. WDYT @kylegach ?

**Comment by yannbf** [comment]
Overall this looks good! But we should definitely test this properly via a canary. I'm triggering a release now so we can test next week!

File: code/renderers/svelte/template/cli/ts-4-9/Button.stories.ts (hunk 3 above)
Comment (kasperpeulen) [reply code/renderers/svelte/template/cli/ts-4-9/Button.stories.ts:2]: We have decided to not yet invest time in this for Svelte, as Svelte 5 will change the way events are handled in svelte.

File: code/renderers/svelte/src/public-types.ts (hunk 5 above)
Comment (kasperpeulen) [reply code/renderers/svelte/src/public-types.ts:44]: I can revert this yes. Was trying around with explicit events for Svelte.

File: code/addons/interactions/template/stories/basics.stories.ts (hunk 8 above)
Comment (kasperpeulen) [reply code/addons/interactions/template/stories/basics.stories.ts:108]: It seems like the paste method is only working with HTTPS in the new version of user event.
It didn't seem like the paste was the point of this test, so I removed it.

File: code/addons/interactions/template/stories/unhandled-errors.stories.ts (hunk 7 above)
Comment (kasperpeulen) [reply code/addons/interactions/template/stories/unhandled-errors.stories.ts:17]: Once chromatic supports unhandled errors, we need to disable snapshots yes (and also we need disable the test runner for this story)

File: code/renderers/react/template/cli/ts-4-9/Button.stories.ts (hunk 6 above)
Comment (kasperpeulen) [reply code/renderers/react/template/cli/ts-4-9/Button.stories.ts:19]: Good point.
Where should we link this to? cc @jonniebigodes @kylegach

Comment (kylegach) [reply code/renderers/react/template/cli/ts-4-9/Button.stories.ts:19]: https://storybook.js.org/docs/8.0/essentials/actions#action-args ?

File: code/addons/interactions/src/components/InteractionsPanel.tsx (hunk 9 above)
Comment (kylegach) [reply code/addons/interactions/src/components/InteractionsPanel.tsx:147]: Yes, agreed. I don't believe that's currently documented anywhere, right?

The upcoming (soon!) parameters API reference would be most appropriate, then. Does the `test` namespace correspond to anything, or is it part of the Storybook core?

Comment (kasperpeulen) [reply code/addons/interactions/src/components/InteractionsPanel.tsx:147]: Part of storybook core, but some parts are only used by the test package, for example, we have those test parameters as well:

image

File: code/renderers/react/template/cli/ts-4-9/Button.stories.ts (hunk 6 above)
Comment (kasperpeulen) [reply code/renderers/react/template/cli/ts-4-9/Button.stories.ts:19]: Great!

Comment (kasperpeulen) [reply code/renderers/react/template/cli/ts-4-9/Button.stories.ts:19]: Shall I also copy this text?
image

Comment (kasperpeulen) [reply code/renderers/react/template/cli/ts-4-9/Button.stories.ts:20]: @kylegach gonna merge them but let me know if you want some tweaking still
